File: Insert Interval/Insert_Interval.py
# 思路二的py代码
# 这样的代码，会导致列表中的对象是共享的，所以还是有一点问题的，但是懒得修改了

# Runtime: 56 ms, faster than 86.15% of Python3 online submissions for Insert Interval.
# Memory Usage: 15.8 MB, less than 100.00% of Python3 online submissions for Insert Interval.

# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[Interval]
        :type newInterval: Interval
        :rtype: List[Interval]
        """
        res = []
        insert = False
        index = 0
        while index != len(intervals):
            if intervals[index].end < newInterval.start:
                res.append(intervals[index])
            elif intervals[index].start > newInterval.end:
                insert = True
                res.append(newInterval)
                res.append(intervals[index])
                break
            else:
                newInterval.start = min(intervals[index].start, newInterval.start)
                newInterval.end = max(intervals[index].end, newInterval.end)
            index += 1
        
        
        if not insert:
            res.append(newInterval)
        else:
            index += 1
            while index != len(intervals):
                res.append(intervals[index])
                index += 1
        return res

File: Insert Interval/Insert_Interval.cpp
// Runtime: 244 ms, faster than 7.56% of C++ online submissions for Insert Interval.
// Memory Usage: 11.2 MB, less than 100.00% of C++ online submissions for Insert Interval.

/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution 
{
public:
    vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) 
    {
        auto iter = intervals.begin();
        for (; iter != intervals.end(); iter++)
        {
            // 位于右边
            if (iter->end < newInterval.start) 
                continue;
            // 位于左边
            else if (iter->start > newInterval.end)
            {
                intervals.insert(iter, newInterval);
                return intervals;
            }
            // 存在交集
            else
            {
                newInterval.start = min(iter->start, newInterval.start);
                newInterval.end = max(iter->end, newInterval.end);
                intervals.erase(iter--);
            }
        }

        intervals.insert(intervals.end(), newInterval);
        return intervals;
    }
};

// 由于vector的随机插入删除效率比较低，因此第二种实现方法是新建一个vector，然后不断在最后进行元素的插入
// Runtime: 16 ms, faster than 100.00% of C++ online submissions for Insert Interval.
// Memory Usage: 11.1 MB, less than 100.00% of C++ online submissions for Insert Interval.

/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */

class Solution 
{
public:
    vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) 
    {
        vector<Interval> res;
        bool insert = false;
        auto iter = intervals.begin();
        for (; iter != intervals.end(); iter++)
        {
            // 位于右边
            if (iter->end < newInterval.start) 
                res.push_back(*iter);
            // 位于左边
            else if (iter->start > newInterval.end)
            {
                insert = true;
                res.push_back(newInterval);
                res.push_back(*iter);
                break;
            }
            // 存在交集
            else
            {
                newInterval.start = min(iter->start, newInterval.start);
                newInterval.end = max(iter->end, newInterval.end);
            }
        }

        if (!insert) 
            res.push_back(newInterval);
        else
        {
            for (iter++; iter != intervals.end(); iter++)
                res.push_back(*iter);
        }
        return res;
    }
};